Worker Node ISO Download with br-dpu Bridge Configuration

szigmon opened the pull request:
**Overview**

This PR adds functionality to easily create and download a discovery ISO for worker nodes with BlueField DPUs. The ISO includes proper network configuration for the br-dpu bridge which is essential for DPU operations.

**Features**

- Adds new Makefile targets:
    - create-day2-cluster: Creates a day2 cluster for worker nodes
    - get-worker-iso: Retrieves the ISO URL for the worker nodes
    - create-cluster-iso: Combines the above operations
-  Implements new shell functions in scripts/cluster.sh:
    - create_day2_cluster(): Creates a day2 cluster with nmstate configuration for the br-dpu bridge
    - download_worker_iso(): Multi-method ISO URL retrieval with fallback mechanisms
- Includes robustness enhancements:
    - Multiple methods to retrieve ISO URL (direct, cluster ID, infraenv approach)
    - Manual override with ISO_URL environment variable
    - Skip URL check with SKIP_ISO_CHECK=true option

**Technical Implementation**

- The day2 cluster uses the "static_network_config" parameter to include nmstate configuration
- The nmstate configuration automatically creates a bridge called "br-dpu" that includes the physical interface
- The bridge is configured for DHCP to automatically get an IP address
- The physical interface is replaced using a REPLACE_INTERFACE placeholder that gets substituted at boot time

**Testing**

The changes were tested with an existing OpenShift cluster with DPF installed. The implementation successfully:

- Creates a day2 cluster with proper configuration
- Provides the ISO URL for manual download
- Includes fallback mechanisms to handle various edge cases

**Documentation**

Documentation changes have been submitted separately to the docs/update-doca-services branch. These changes include:

- Detailed instructions for the new workflow
- Explanation of the br-dpu bridge configuration
- Troubleshooting tips for various ISO download scenarios
- iDRAC-specific instructions for booting worker nodes

tsorya commented on scripts/cluster.sh:
```diff
@@ -215,6 +215,118 @@ function download_iso() {
+            -P public_key="${SSH_KEY}" \
+            -P day2=true \
+            -P static_network_config="interfaces:
+  - name: br-dpu
```
You don't need this part, there is no point to configure br-dpu before joining to cluster if you don't setup static ip.

szigmon replied:
Both options can work, but since we have MachineConfig logic already, let's skip it.

tsorya commented on scripts/cluster.sh:
```diff
@@ -215,6 +215,156 @@ function download_iso() {
+            
+            # Try direct infraenv approach
+            if [ -n "${cluster_id}" ]; then
+                iso_url=$(aicli info infraenv ${cluster_id} 2>/dev/null | grep -oP 'download_url: \K.*' 2>/dev/null || echo "")
```
Sure this works? How infraenv id and cluster id correlates?

szigmon commented on scripts/cluster.sh:
```diff
@@ -215,6 +215,118 @@ function download_iso() {
+        # The REPLACE_INTERFACE placeholder will be automatically replaced during node boot
+        # with the interface that has the default route
+        aicli create cluster \
+            -P openshift_version="${OPENSHIFT_VERSION}" \
```
I think yes, from my understanding these are the required ones at least:
1. openshift_version: Required to ensure the worker nodes use the same version as the main cluster
2. base_dns_domain: Required for proper DNS resolution in the cluster
3. pull_secret: Required for accessing OpenShift images
4. public_key: Required for SSH access to the hosts
5. day2=true: Essential - indicates this is a day2 operation to add hosts to existing cluster

szigmon commented on scripts/cluster.sh:
```diff
@@ -215,6 +215,118 @@ function download_iso() {
+    fi
+}
+
+function download_worker_iso() {
```
The separate function is needed because the use case is different. Unlike the main cluster ISO which we download directly to the automation machine, the worker node ISO is intended to be accessed by the user's workstation for uploading to iDRAC

This function:

1. Provides a URL for manual download instead of downloading directly
2. Includes multiple retrieval methods and fallbacks since URL fetching can be challenging
3. Has custom error handling to guide users through manual alternatives
4. Performs version verification to ensure consistency with the main cluster

A direct download isn't practical since users need to transfer the ISO to their local workstation to upload it to iDRAC virtual media."

tsorya commented on docs/prerequisites.md:
```diff
@@ -0,0 +1 @@
+ 
```
why it is required empty?

tsorya commented on docs/dpu-provisioning.md:
```diff
@@ -0,0 +1,782 @@
+# DPU Provisioning
```
Maybe lets have it as separate pr?

szigmon commented:
**PR Summary:** Improve Worker Node ISO Management

This PR improves the ISO management for BlueField DPU worker nodes:

**Changes Made:**

1. Added a unified get_iso function with an intuitive parameter structure
2. Simplified worker node ISO retrieval with clear command options
3. Added detailed documentation for BlueField DPU worker node integration
4. Fixed infraenv lookup for more reliable ISO URL retrieval

**Key Features:**

- Simple commands for worker node ISO management:
    - make get-worker-iso - Get ISO URL for worker nodes
    - make create-cluster-iso - Create day2 cluster and get worker ISO URL
- Improved error handling with helpful user guidance
- Automatic version consistency checks between main and day2 clusters

tsorya commented on Makefile:
```diff
@@ -108,18 +111,30 @@ install-hypershift:
+create-day2-cluster:
+	@$(CLUSTER_SCRIPT) create-day2-cluster
+
+get-worker-iso:
```
It probably should depend on create-day2-cluster and no point to have create-cluster-iso

tsorya commented on Makefile:
```diff
@@ -44,6 +44,9 @@ update-paths:
 download-iso:
 	@$(CLUSTER_SCRIPT) download-iso
 
+get-iso:
```
why this one is required?

tsorya commented on scripts/cluster.sh:
```diff
@@ -215,6 +215,329 @@ function download_iso() {
+    fi
+    
+    # Color codes for output formatting
+    local YELLOW='\033[1;33m'
```
I don't think we need that much coloring, just leave it default

tsorya commented on scripts/cluster.sh:
```diff
@@ -215,6 +215,329 @@ function download_iso() {
+    local day2_cluster_name="${cluster_name}-day2"
+    
+    # Export for use in other functions
+    export CLUSTER_NAME="${cluster_name}"
```
remove this one, you have all data coming from env.sh , no point to have those here

tsorya replied:
Why to export  CLUSTER_NAME from local var?

tsorya commented on scripts/cluster.sh:
```diff
@@ -215,6 +215,329 @@ function download_iso() {
+    
+    log "INFO" "ISO request: cluster=${cluster_name}, type=${node_type}, action=${action}, path=${download_path}, iso_type=${iso_type}"
+    
+    if [[ "${node_type}" != "master" && "${node_type}" != "worker" ]]; then
```
Node type should not matter, just download iso from cluster name, you can provide day2 boolean probably and no need for this check

tsorya commented on scripts/cluster.sh:
```diff
@@ -215,6 +215,329 @@ function download_iso() {
+                local console_url="https://console.redhat.com/openshift/assisted-installer/clusters"
+                
+                # Display URL with formatting
+                echo ""
```
why all this formatting is required? This creates tons of text for which exactly purpose? Lets just pring simple text without all the alternatives. Simplify it please

tsorya commented on scripts/cluster.sh:
```diff
@@ -215,6 +215,329 @@ function download_iso() {
+        fi
+    else
+        # Master nodes use the main cluster for ISO
+        if [ "${action}" = "download" ]; then
```
didn't we talk to use already existing function?

tsorya commented on Makefile:
```diff
@@ -108,18 +111,35 @@ install-hypershift:
+	@$(CLUSTER_SCRIPT) get-worker-iso
+
+create-cluster-iso:
+	@if [ -z "$(ISO_TYPE)" ]; then \
```
do it in script an not here, the idea was to remove logic from makefile

tsorya commented on scripts/cluster.sh:
```diff
@@ -215,6 +215,220 @@ function download_iso() {
+    fi
+    
+    # For worker nodes, we need a day2 cluster
+    if [ "${node_type}" = "worker" ]; then
```
I don't like to have node type at all, this should be boolean , day2=true/false

tsorya commented on scripts/cluster.sh:
```diff
@@ -215,6 +215,220 @@ function download_iso() {
+            
+            # Fallback to direct method if infraenv failed
+            if [ -z "${iso_url}" ]; then
+                iso_url=$(aicli info iso "${day2_cluster_name}" 2>/dev/null | grep -i "url:" | awk '{print $2}')
```
This is totally enough to download iso, please remove all other options. No need to fallback

tsorya commented on scripts/cluster.sh:
```diff
@@ -215,6 +215,81 @@ function download_iso() {
+        
+        # Get original cluster version
+        local original_version=$(aicli info cluster ${CLUSTER_NAME} -f openshift_version -v 2>/dev/null || echo "")
+        [ -z "${original_version}" ] && { log "ERROR" "Could not find cluster ${CLUSTER_NAME}"; exit 1; }
```
I don't think this is interesting to set original version and base_dns_domain

szigmon commented:
**Overview**
This PR significantly improves and simplifies the worker node ISO functionality for OpenShift DPF deployments. The implementation follows strict DRY principles with a single, unified function for all ISO operations that is easy to maintain and extend.

**Key Improvements**

1. Single, Unified Function: Consolidated all ISO operations into a single get_iso function that handles both worker and master node operations
    - Eliminated redundant wrapper functions and duplicate code
    - Function parameters handle all use cases: cluster_name, node_type, action, download_path, and iso_type

2. Token-Based URL Handling: Added robust support for token-based URLs required for authenticated ISO downloads
    - Used InfraEnv approach as primary method for retrieving JWT-authenticated URLs
    - Preserved token URL format while allowing ISO type selection (minimal/full)

3. Simplified Fallback Logic: Implemented clean fallback mechanism when direct URL retrieval fails

    - Falls back to console.redhat.com UI URL as a reliable alternative
    - Clear error messages with helpful instructions

4. Improved Default Values: Set sensible defaults for all parameters
      - Default ISO type is now explicitly "minimal"
      - Removed dependency on environment variables for defaults
      - Preserves backward compatibility with Makefiles and existing command patterns

5. Documentation Updates: Simplified usage instructions in documentation
    - Clearer, more concise steps for getting worker ISO URLs
    - Straightforward troubleshooting guidance

**Code Quality Improvements**

- Reduced Code Size: Cut the implementation by ~70% while maintaining all functionality
- Better Organization: Cleaner control flow and error handling
- Less Verbosity: Reduced excessive logging while keeping essential messages
- No Duplication: Strict adherence to DRY principles


**Response to Review Comments**
- Removed unnecessary static network configuration for br-dpu bridge as requested (using MachineConfig instead)
- Simplified ISO URL retrieval to use the most reliable method (InfraEnv approach for token-based URLs)
- Eliminated all redundant fallback methods while keeping one reliable fallback for edge cases
- Focused on consistent error messaging with clear user guidance

**Testing**

The implementation has been tested with:

- Both minimal and full ISO types
- Token-based URL retrieval scenarios
- Direct command-line usage and Makefile targets
- Fallback to UI URL when needed


**Usage**
The primary methods for getting worker ISO URLs remain unchanged:
`# Get minimal ISO URL (default)
make get-worker-iso

# Get full ISO URL
ISO_TYPE=full make get-worker-iso`

tsorya commented on test-worker-iso.sh:
```diff
@@ -0,0 +1,131 @@
+#!/bin/bash
+# test-worker-iso.sh - Test worker ISO download functionality
```
remove this one
